ts/private/ts_project_worker.js

```diff
@@ -0,0 +1,168 @@
+            },
+            watchDirectory: (path, callback, recursive, options) => {
+                // TODO: Figure out what to do with directory watchers
+                // tsc seem to be using them for watching the typeroots, node_modules tree, current dir
+            }
+        }
+  
+        const host = ts.createWatchCompilerHost(
+            tsconfigPath,
+            parsedArgs.options,
+            sys,
+            ts.createSemanticDiagnosticsBuilderProgram,
+            noop,
+            noop
+        )
+
+        lastRequestArgHash = args.join(' ')
+        program = ts.createWatchProgram(host)
+    }
+
+    return program
+}
+
+async function emit(args, inputs, once = false) {
+    lastRequestTimestamp = Date.now()
+
+
+    if (previousInputs) {
+        console.time(`invalidate`);
+        for (const input of Object.keys(previousInputs)) {
+            if (!inputs[input]) {
+                invalidate.get(input)?.(ts.FileWatcherEventKind.Deleted)
+            }
+        }
+        for (const [input, digest] of Object.entries(inputs)) {
+            if (!(input in previousInputs)) {
+                invalidate.get(input)?.(ts.FileWatcherEventKind.Created)
+            } else if (previousInputs[input] != digest) {
+                invalidate.get(input)?.(ts.FileWatcherEventKind.Changed)
+            }
+        }
+        console.timeEnd("invalidate")
+    }
+ 
+
+    console.time("getProgram")
+    // NOTE: Always get the program after the invalidation logic above as watcher program could
+    // swap the program with new one based on the changes
+    const program = getProgram(args).getProgram()
+    console.timeEnd("getProgram")
+
+    const cancellationToken = {
+        isCancellationRequested: function (timestamp) {
+            return timestamp !== lastRequestTimestamp
+        }.bind(null, lastRequestTimestamp),
+        throwIfCancellationRequested: function (timestamp) {
+            if (timestamp !== lastRequestTimestamp) {
+                throw new ts.OperationCanceledException()
+            }
+        }.bind(null, lastRequestTimestamp),
+    }
+
+    console.time("emit")
+    const result = program.emit(undefined, undefined, cancellationToken)
+    console.timeEnd("emit")
+    console.time("diagnostics")
+    const diagnostics = ts.getPreEmitDiagnostics(
+        program,
+        undefined,
+        cancellationToken
+    )
+    console.timeEnd("diagnostics")
+    const succeded =
+        !result.emitSkipped &&
+        result?.diagnostics.length === 0 &&
+        diagnostics.length == 0
+
+    if (!succeded) {
+        printDiagnostics(diagnostics)
+    }
+
+    previousInputs = inputs
+
+    return succeded
+}
+
+function main() {
+    if (worker.runAsWorker(process.argv)) {
+        worker.log(`Running ${MNEMONIC} as a Bazel worker`)
+        worker.runWorkerLoop(emit)
+    } else {
+        worker.log(`Running ${MNEMONIC} as a standalone process`)
```

Comment:
This code path is broken looks like. Something to do with parsing the args from the params file:
```
bazel build //examples/worker:ts --spawn_strategy=darwin-sandbox

ERROR: /Users/greg/aspect/rules/js/rules_ts/examples/worker/BUILD.bazel:14:11: Compiling TypeScript project (worker mode) //examples/worker:ts [tsc -p examples/worker/tsconfig.json] failed: (Exit 1): _tsc_worker_launcher.sh failed: error executing command bazel-out/darwin-opt-exec-2B5CBBC6/bin/external/npm_typescript/_tsc_worker_launcher.sh @bazel-out/darwin-fastbuild/bin/examples/worker/foo.js-0.params


Use --sandbox_debug to see verbose messages from the sandbox
Running TsProject as a standalone process
Started a new process to perform this action. Your build might be misconfigured, try 
 build --strategy=TsProject=worker
node:internal/fs/utils:344
    throw err;
    ^

Error: ENOENT: no such file or directory, open 'bazel-out/darwin-fastbuild/bin/examples/worker/foo.js-0.params'
    at Object.openSync (node:fs:585:3)
    at Object.readFileSync (node:fs:453:35)
```